src/v/config/property.h

```diff
@@ -123,6 +132,152 @@ class property : public base_property {
     };
 };
 
+template<typename T>
+concept has_type_name = requires(T x) {
+    x.type_name();
+};
+
+template<typename T>
+concept is_collection = requires(T x) {
+    typename T::value_type;
+    {x.size()};
+};
+
+template<typename T>
+consteval std::string_view property_type_name_inner() {
+    using type = std::decay_t<T>;
+    if constexpr (std::is_same_v<type, ss::sstring>) {
+        // String check must come before is_collection check
+        return "string";
+    } else if constexpr (std::is_same_v<type, bool>) {
+        // boolean check must come before is_integral check
+        return "boolean";
+    } else if constexpr (reflection::is_std_optional_v<type>) {
+        return property_type_name_inner<typename type::value_type>();
+    } else if constexpr (is_collection<type>) {
+        return property_type_name_inner<typename type::value_type>();
+    } else if constexpr (has_type_name<type>) {
+        return type::type_name();
+    } else if constexpr (std::is_same_v<type, model::compression>) {
+        return "string";
+    } else if constexpr (std::is_same_v<type, model::timestamp_type>) {
+        return "string";
+    } else if constexpr (std::is_same_v<type, model::cleanup_policy_bitflags>) {
+        return "string";
+    } else if constexpr (std::
+                           is_same_v<type, model::violation_recovery_policy>) {
+        return "string";
+    } else if constexpr (std::is_same_v<type, config::data_directory_path>) {
+        return "string";
+    } else if constexpr (std::is_same_v<type, model::node_id>) {
+        return "integer";
+    } else if constexpr (std::is_same_v<type, std::chrono::seconds>) {
+        return "integer";
+    } else if constexpr (std::is_same_v<type, std::chrono::milliseconds>) {
+        return "integer";
+    } else if constexpr (std::is_same_v<type, seed_server>) {
+        return "seed_server";
+    } else if constexpr (std::is_same_v<type, unresolved_address>) {
+        return "unresolved_address";
+    } else if constexpr (std::is_same_v<type, tls_config>) {
+        return "tls_config";
+    } else if constexpr (std::is_same_v<type, endpoint_tls_config>) {
+        return "endpoint_tls_config";
+    } else if constexpr (std::is_same_v<type, model::broker_endpoint>) {
+        return "broker_endpoint";
+    } else if constexpr (std::is_floating_point_v<type>) {
+        return "number";
+    } else if constexpr (std::is_integral_v<type>) {
+        return "integer";
+    }
+
+    // Never used, the `property_type_name` wrapper requires that
+    // this value is not returned when instantiating the function for
+    // any type.
+    return "";
```

Comment:
Does it make sense to do:

```
template<typename T>
struct dependent_false : std::false_type {};

...
} else {
    static_assert(details::dependent_false<T>::value, "some useful message");
}
```

then I think you could avoid `requires(property_type_name_inner<T>() != "") consteval std::string_view`?

Reply:
When I typed out the example, I forgot the static_assert wasn't dependent, this is nice.

Reply:
That is nicer -- done.